TRUNK-6155: Upgrade 2.7 Platform to Liquibase 4.8.0

k4pran opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'TRUNK-5111: Replace use of deprecated isVoided' -->
<!--- 'TRUNK-JiraIssueNumber: JiraIssueTitle' -->
## Description of what I changed
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->

Upgraded liquibase core and liquibase plugin to version 4.8.0. See the jira ticket below to review attached logs from manual testing.

There was a breaking change in version 4.7.0. This isn't shown in the github release notes but you can find them here - https://docs.liquibase.com/start/release-notes/home.html. See below

----------------------------------
**Breaking Changes**

[[PR#2228](https://github.com/liquibase/liquibase/pull/2228)] [[Nathan Voxland](https://github.com/nvoxland)] Handle the fact that FK names are not always unique. Fixes [PR#2227](https://github.com/liquibase/liquibase/issues/2227). When using the foreignKeyConstraintExists precondition, you must now specify foreignKeyTableName. See also: [PR#2565](https://github.com/liquibase/liquibase/pull/2565) fixing [#2389](https://github.com/liquibase/liquibase/issues/2389). in 4.11.0.
----------------------------------

I added the table name to the constraints to fix this breaking change.

## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first! -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it! -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/TRUNK-6155

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

mherman22 reviewed (changes requested):
@k4pran thanks for your work here. Take note of;-
- You added the wrong ticket ID, do you mind correcting that?
- The build is failing probably because of the big leap in versions, take a look at those please since alot could have happened in between the two versions.

k4pran commented:
> @k4pran thanks for your work here. Take note of;-
> 
> * You added the wrong ticket ID, do you mind correcting that?
> * The build is failing probably because of the big leap in versions, take a look at those please since alot could have happened in between the two versions.

My mistake I fixed the ticket number. I did some testing locally for the version update but seems the build is picking up some issues, will convert this PR to a draft while I look into it

k4pran commented:
> @k4pran thanks for your work here. Take note of;-
> 
> * You added the wrong ticket ID, do you mind correcting that?
> * The build is failing probably because of the big leap in versions, take a look at those please since alot could have happened in between the two versions.

I fixed the issue and updated the description of this PR, there was a breaking change that wasn't documented in the github release notes. Seems to be building now. I'm not sure why it was passing locally though. Seemed to pass running mvn clean install, but failed when I ran the individual tests.

dkayiwa commented:
> I'm not sure why it was passing locally though. Seemed to pass running mvn clean install, but failed when I ran the individual tests.

For the integration tests which failed, you would need to run `mvn test -Pskip-default-test -Pintegration-test`

ibacher commented on api/src/main/resources/org/openmrs/liquibase/updates/liquibase-update-to-latest-2.1.x.xml:
```diff
@@ -42,14 +42,14 @@
 	<changeSet author="vshankar" id="201609171146-1">
 		<preConditions onFail="MARK_RAN">
-			<foreignKeyConstraintExists foreignKeyName="member_patient"/>
+			<foreignKeyConstraintExists foreignKeyTableName="cohort_member" foreignKeyName="member_patient"/>
```
@k4pran Is this required or just a hedge against the future? If it's required, we just need to ensure that we make similar updates to the modules as well.

k4pran replied:
> @k4pran Is this required or just a hedge against the future? If it's required, we just need to ensure that we make similar updates to the modules as well.

This would be required for integration tests to pass as some of them run through all the update xmls, liquibase now require the table to be referenced when setting up these constraints.

I didn’t realise other modules have update xmls but yes I think we should update them all. Do you know what modules will have them?

I guess another question, are there cases where clients using openmrs will have their own update xml changesets that we don’t have access to? I am just thinking if they have taken a snapshot that don’t have the foreignKeyTableName?

I also added a link to the breaking change in the PR description if you want to read a bit more about it

ibacher replied:
> Do you know what modules will have them?

Well, at least the modules turned up [by this search](https://github.com/search?q=org%3Aopenmrs+foreignKeyConstraintExists&type=code)

> I guess another question, are there cases where clients using openmrs will have their own update xml changesets that we don’t have access to?

That's a real situation, but we can easily tell people who run into the issue how to fix it. Most modules (as opposed to core) rely on hand-maintained Liquibase scripts rather than snapshots.

> I also added a link to the breaking change in the PR description if you want to read a bit more about it

Ah, I was under the impression that that change was is Liquibase 4.11, not 4.7. I see.

k4pran replied:
Yea liquibase 4.11 has a few additional breaking changes and also a few code changes we will need. I haven't looked deeply into the additional breaking changes. I will update the constraints for those other repos too. Should it be new tickets for each repo?

ibacher replied:
@k4pran Yeah, most modules have their own separate Jira project, so we should stick to that.

gracepotma commented:
Thank you **so much** @k4pran for your work on this!! We're so thankful for your involvement and contributions. 💐 🏆 It's _especially_ valuable and appreciated when folks help us with this kind of platform work. Can't thank you enough!
